src/coap_openssl.c

```diff
@@ -492,138 +510,724 @@ int coap_dtls_context_set_psk(coap_context_t *ctx,
       }
     }
     else {
-      coap_log(LOG_ERR, "*** coap_dtls_context_set_pki: DTLS: No Server Certificate defined\n");
+      coap_log(LOG_ERR,
+           "*** setup_pki: (D)TLS: No Server Private Key defined\n");
+      return 0;
     }
-    if (setup_data->private_key && setup_data->private_key[0]) {
-      if (!(SSL_CTX_use_PrivateKey_file(context->dtls.ctx, setup_data->private_key, SSL_FILETYPE_PEM))) {
-        coap_log(LOG_WARNING, "*** coap_dtls_context_set_pki: DTLS: %s: Unable to configure Server Private Key\n", setup_data->private_key);
+
+    if (setup_data->key.pem.ca_file && setup_data->key.pem.ca_file[0]) {
+      STACK_OF(X509_NAME) *cert_names;
+      X509_STORE *st;
+      long e;
+      BIO *in;
+      X509 *x = NULL;
+      char *rw_var = NULL;
+      cert_names = SSL_load_client_CA_file(setup_data->key.pem.ca_file);
+      if (cert_names != NULL)
+        SSL_CTX_set_client_CA_list(ctx, cert_names);
+      else {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure client CA File\n",
+                  setup_data->key.pem.ca_file);
         return 0;
       }
-    }
-    else if (setup_data->asn1_private_key && setup_data->asn1_private_key_len > 0) {
-      int pkey_type;
-      switch (setup_data->asn1_private_key_type) {
-      case COAP_ASN1_PKEY_NONE:
-        pkey_type = EVP_PKEY_NONE;
+      st = SSL_CTX_get_cert_store(ctx);
+      in = BIO_new(BIO_s_file());
+      /* Need to do this to not get a compiler warning about const parameters */
+      memcpy(&rw_var, &setup_data->key.pem.ca_file, sizeof (rw_var));
+      if (!BIO_read_filename(in, rw_var)) {
+        BIO_free(in);
+        X509_free(x);
         break;
-      case COAP_ASN1_PKEY_RSA:
-        pkey_type = EVP_PKEY_RSA;
-        break;
-      case COAP_ASN1_PKEY_RSA2:
-        pkey_type = EVP_PKEY_RSA2;
-        break;
-      case COAP_ASN1_PKEY_DSA:
-        pkey_type = EVP_PKEY_DSA;
-        break;
-      case COAP_ASN1_PKEY_DSA1:
-        pkey_type = EVP_PKEY_DSA1;
-        break;
-      case COAP_ASN1_PKEY_DSA2:
-        pkey_type = EVP_PKEY_DSA2;
-        break;
-      case COAP_ASN1_PKEY_DSA3:
-        pkey_type = EVP_PKEY_DSA3;
-        break;
-      case COAP_ASN1_PKEY_DSA4:
-        pkey_type = EVP_PKEY_DSA4;
-        break;
-      case COAP_ASN1_PKEY_DH:
-        pkey_type = EVP_PKEY_DH;
-        break;
-      case COAP_ASN1_PKEY_DHX:
-        pkey_type = EVP_PKEY_DHX;
-        break;
-      case COAP_ASN1_PKEY_EC:
-        pkey_type = EVP_PKEY_EC;
-        break;
-      case COAP_ASN1_PKEY_HMAC:
-        pkey_type = EVP_PKEY_HMAC;
-        break;
-      case COAP_ASN1_PKEY_CMAC:
-        pkey_type = EVP_PKEY_CMAC;
-        break;
-      case COAP_ASN1_PKEY_TLS1_PRF:
-        pkey_type = EVP_PKEY_TLS1_PRF;
-        break;
-      case COAP_ASN1_PKEY_HKDF:
-        pkey_type = EVP_PKEY_HKDF;
-        break;
-      default:
+      }
+
+      /* Flush out existing errors */
+      while ((e = ERR_get_error()) != 0) { }
```

Comment:
For better readability, please move the closing parenthesis to the next line.

Reply:
Fixed